sdk/tests/deqp/framework/referencerenderer/rrRenderer.js

```diff
@@ -550,6 +522,372 @@ rrRenderer.clipTest = function(x, y, z, rect) {
+
+    // Positions in fixed-point coordinates
+    var x0 = rrRenderer.toSubpixelCoord(v0[0]);
+    var y0 = rrRenderer.toSubpixelCoord(v0[1]);
+    var x1 = rrRenderer.toSubpixelCoord(v1[0]);
+    var y1 = rrRenderer.toSubpixelCoord(v1[1]);
+    var x2 = rrRenderer.toSubpixelCoord(v2[0]);
+    var y2 = rrRenderer.toSubpixelCoord(v2[1]);
+
+    // Initialize edge functions
+    if (this.m_winding == rrRenderState.Winding.CCW) {
+        this.m_edge01 = rrRenderer.initEdgeCCW(this.m_horizontalFill, this.m_verticalFill, x0, y0, x1, y1);
+        this.m_edge12 = rrRenderer.initEdgeCCW(this.m_horizontalFill, this.m_verticalFill, x1, y1, x2, y2);
+        this.m_edge20 = rrRenderer.initEdgeCCW(this.m_horizontalFill, this.m_verticalFill, x2, y2, x0, y0);
+    } else {
+        // Reverse edges
+        this.m_edge01 = rrRenderer.initEdgeCCW(this.m_horizontalFill, this.m_verticalFill, x1, y1, x0, y0);
+        this.m_edge12 = rrRenderer.initEdgeCCW(this.m_horizontalFill, this.m_verticalFill, x2, y2, x1, y1);
+        this.m_edge20 = rrRenderer.initEdgeCCW(this.m_horizontalFill, this.m_verticalFill, x0, y0, x2, y2);
+    }
+
+    // Determine face
+    var s = this.m_edge01.evaluateEdge(x2, y2);
+    var positiveArea = (this.m_winding == rrRenderState.Winding.CCW ) ? s > 0 : s < 0;
+    this.m_face = positiveArea ? rrDefs.FaceType.FACETYPE_FRONT : rrDefs.FaceType.FACETYPE_BACK;
+    if (!positiveArea) {
+        // Reverse edges so that we can use CCW area tests & interpolation
+        this.m_edge01.reverseEdge();
+        this.m_edge12.reverseEdge();
+        this.m_edge20.reverseEdge();
+    }
+
+    // Bounding box
+    var minX = Math.min(x0, x1, x2);
+    var maxX = Math.max(x0, x1, x2);
+    var minY = Math.min(y0, y1, y2);
+    var maxY = Math.max(y0, y1, y2);
+
+    this.m_bboxMin = [];
+    this.m_bboxMax = [];
+    this.m_bboxMin[0] = rrRenderer.floorSubpixelToPixelCoord(minX, this.m_horizontalFill == rrRenderState.HorizontalFill.LEFT);
+    this.m_bboxMin[1] = rrRenderer.floorSubpixelToPixelCoord(minY, this.m_verticalFill == rrRenderState.VerticalFill.BOTTOM);
+    this.m_bboxMax[0] = rrRenderer.ceilSubpixelToPixelCoord(maxX, this.m_horizontalFill == rrRenderState.HorizontalFill.RIGHT);
+    this.m_bboxMax[1] = rrRenderer.ceilSubpixelToPixelCoord(maxY, this.m_verticalFill == rrRenderState.VerticalFill.TOP);
+
+    // Clamp to viewport
+    var wX0 = this.m_viewport.rect.left;
+    var wY0 = this.m_viewport.rect.bottom;
+    var wX1 = wX0 + this.m_viewport.rect.width - 1;
+    var wY1 = wY0 + this.m_viewport.rect.height - 1;
+
+    this.m_bboxMin[0] = deMath.clamp(this.m_bboxMin[0], wX0, wX1);
+    this.m_bboxMin[1] = deMath.clamp(this.m_bboxMin[1], wY0, wY1);
+    this.m_bboxMax[0] = deMath.clamp(this.m_bboxMax[0], wX0, wX1);
+    this.m_bboxMax[1] = deMath.clamp(this.m_bboxMax[1], wY0, wY1);
+
+    this.m_curPos = [this.m_bboxMin[0], this.m_bboxMin[1]];
+};
+
+rrRenderer.triangleRasterizer.prototype.rasterize = function() {
```

Comment:
C++ code:
https://android.googlesource.com/platform/external/deqp/+/master/framework/referencerenderer/rrRasterizer.cpp#531
There are some varyings from C++ code due to the difference of drawTriangles between JS and C++ code